### PR TITLE
Cap dataloader workers for tiny datasets

### PR DESCRIPTION
**Pull request by glenn-jocher**
## Summary
- Cap dataloader workers from the final loader batch count after sampler construction, while preserving existing global `drop_last` semantics.
- Run single-batch loaders in-process so tiny training jobs do not spawn persistent worker pools that can outlive a wedged trainer and keep CUDA contexts busy.
- Preserve PyTorch DataLoader validation for empty datasets instead of introducing a worker-cap `ZeroDivisionError`.
- Add regression coverage for single-process tiny loaders, distributed sampler/drop-last behavior, and empty-dataset validation.
- Bump package version to `8.4.88` for Portal adoption.

Deleted: excess dataloader worker processes for single-batch loaders; custom test dataset scaffolding in favor of built-in `range` datasets.

Rule 2 justification: net-positive because the bugfix needs regression coverage for single-process, distributed sampler, and empty-dataset semantics; there was no existing focused test for this worker-count path.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Improves dataloader worker handling for small datasets, reducing unnecessary multiprocessing overhead and potential stalls 🚀

### 📊 Key Changes
- Added logic in `build_dataloader()` to cap worker processes based on the actual number of batches created.
- Single-batch dataloaders now run in-process with `num_workers=0` to avoid unnecessary persistent worker pools.
- Worker calculation now accounts for distributed sampling, ensuring correct behavior in multi-GPU or distributed training setups.
- Preserves `drop_last` behavior while avoiding edge-case issues with tiny or empty datasets.
- Added new tests covering:
  - Tiny datasets with one or two batches ✅
  - Distributed sampling with `drop_last=True` 🌐
  - Empty dataset validation behavior 🧪
- Bumped package version from `8.4.87` to `8.4.88`.

### 🎯 Purpose & Impact
- Reduces overhead for small datasets by avoiding extra worker processes that provide no benefit ⚡
- Helps prevent stalls or resource contention caused by persistent DataLoader workers holding CUDA context.
- Improves reliability for training and validation on tiny datasets, quick tests, and edge-case workflows.
- Maintains correct behavior for distributed training users, minimizing risk of regressions.
- Users should see smoother, more efficient dataloader behavior without needing to change their code 🎉

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @glenn-jocher, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This automated message confirms your PR has been received, and an Ultralytics engineer will assist with review. To ensure a seamless integration of your dataloader worker cap update, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/data/build.py [hunk 1]
```diff
@@ -340,8 +340,11 @@ def build_dataloader(
         >>> dataloader = build_dataloader(dataset, batch=16, workers=4, shuffle=True)
     """
     batch = min(batch, len(dataset))
+    batches = math.ceil(len(dataset) / max(batch, 1))
```
Comment (UltralyticsAssistant) [inline ultralytics/data/build.py]: 📝 **LOW**: `batches` is used as the cap for useful worker processes, but this calculation ignores `drop_last`. For example, `len(dataset)=5`, `batch=4`, `drop_last=True` yields only one emitted batch while this computes two, so a single-batch loader can still create workers and miss the tiny-dataset in-process behavior this PR is adding. Please account for `drop_last` when computing the useful batch count.

**Suggested change:**
```suggestion
    batches = max(len(dataset) // max(batch, 1) if drop_last else math.ceil(len(dataset) / max(batch, 1)), 1)
```

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The change is focused and covered by a small regression test. One edge case needs attention: the worker cap should use the actual number of emitted batches when `drop_last=True`, otherwise some single-batch loaders can still spawn workers.

💬 Posted 1 inline comment

**Comment by UltralyticsAssistant** [comment]
Merged — thank you @glenn-jocher! 🚀

“Great things are done by a series of small things brought together.” — Vincent van Gogh

This PR is a great example of that principle in action: refining `build_dataloader()` to better handle tiny, single-batch, empty, and distributed datasets may seem like a small internal improvement, but it meaningfully reduces overhead, avoids unnecessary worker pools, and improves reliability for training and validation edge cases.

The added test coverage around small datasets, distributed `drop_last=True`, and empty dataset behavior is especially valuable for keeping Ultralytics robust across real-world workflows. Thanks for the thoughtful improvement and continued attention to performance and stability! 🙌
